app/models/candidate_interface/gcse_qualification_details_form.rb

```diff
@@ -17,16 +18,39 @@ def self.build_from_qualification(qualification)
     def save_base
       return false unless valid?
 
-      qualification = ApplicationQualification.find(qualification_id)
-
       qualification.update(grade: grade, award_year: award_year)
     end
 
   private
-
     def validate_award_year_is_date
       valid_award_year = award_year.match(/^[1-9]\d{3}$/)
       errors.add(:award_year, :invalid) unless valid_award_year
     end
+
+    def validate_grade_format
+      return if new_record? || qualification.qualification_type.nil?
+
+      qualification_rexp = invalid_grades[qualification.qualification_type.to_sym]
+
+      if qualification_rexp && grade.match(qualification_rexp)
+        errors.add(:grade, :invalid)
+      end
+    end
+
+    def qualification
+      @qualification ||= ApplicationQualification.find(qualification_id)
+    end
+
+    def invalid_grades
+      {
+        gcse: /[^1-9A-GU\*]/i,
```

Comment:
Wouldn't this permit `B*`, `U*` etc?

Reply:
that's true, I'll see if this fix is not too complex to implement.

Reply:
That's right, anyway, our first implementation of the validation quite permissive.

Once we have a system to log the wrong attempts (described in https://trello.com/c/zlfQ2il2/358-keep-track-of-validation-errors-in-the-application-form) we'll be able to be more precise on the validations.